tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php

```diff
@@ -0,0 +1,207 @@
+                }
+            }
+
+            $restrictionGroups = [];
+            if (isset($this->productRestrictionGroupsByReference[$cartRuleReference])) {
+                $restrictionGroups = $this->productRestrictionGroupsByReference[$cartRuleReference];
+            }
+
+            $restrictionGroups[] = new RestrictionRuleGroup($quantity, $rules);
+            $this->productRestrictionGroupsByReference[$cartRuleReference] = $restrictionGroups;
+        } catch (CartRuleConstraintException $e) {
+            $this->setLastException($e);
+        }
+    }
+
+    /**
+     * @When I restrict cart rules for :cartRuleReference providing non-existing cart rules
+     *
+     * @param string $cartRuleReference
+     *
+     * @return void
+     */
+    public function restrictCartRulesProvidingNonExistingIds(string $cartRuleReference): void
+    {
+        $command = $this->getRestrictionsCommand($cartRuleReference);
+        $command->setRestrictedCartRuleIds([self::NON_EXISTING_CART_RULE_ID]);
+
+        try {
+            $this->getCommandBus()->handle($command);
+        } catch (CartRuleNotFoundException $e) {
+            $this->setLastException($e);
+        }
+    }
+
+    /**
+     * @When I clear all product restrictions for cart rule :cartRuleReference
+     *
+     * @param string $cartRuleReference
+     *
+     * @return void
+     */
+    public function clearProductRestrictionRules(string $cartRuleReference): void
+    {
+        $command = $this->getRestrictionsCommand($cartRuleReference);
+        $command->setProductRestrictionRuleGroups([]);
+
+        $this->restrictionCommandsByReference[$cartRuleReference] = $command;
+        $this->productRestrictionGroupsByReference[$cartRuleReference] = [];
+    }
+
+    /**
+     * @When I clear cart rule combination restrictions for cart rule :cartRuleReference
+     *
+     * @param string $cartRuleReference
+     *
+     * @return void
+     */
+    public function clearCartRuleCombinationRestrictions(string $cartRuleReference): void
+    {
+        $command = $this->getRestrictionsCommand($cartRuleReference);
+        $command->setRestrictedCartRuleIds([]);
+
+        $this->restrictionCommandsByReference[$cartRuleReference] = $command;
+    }
+
+    /**
+     * @When I save all the restrictions for cart rule :cartRuleReference
+     *
+     * @param string $cartRuleReference
+     *
+     * @return void
+     */
+    public function saveRestrictionRules(string $cartRuleReference): void
+    {
+        $command = $this->getRestrictionsCommand($cartRuleReference);
+
+        if (isset($this->productRestrictionGroupsByReference[$cartRuleReference])) {
+            // set the restrictions that (if) were added step by step before
+            $command->setProductRestrictionRuleGroups($this->productRestrictionGroupsByReference[$cartRuleReference]);
+            unset($this->productRestrictionGroupsByReference[$cartRuleReference]);
```

Comment:
Shouldn't `$this->restrictionCommandsByReference` also be cleaned once it's saved? To make sure no values from previous scenarios can remain in case a new scenario starts on the same cart rule